D2k damage calculations

penev92 opened the pull request:
This implements the fourth and final point on the TODO list in #17972 by adding a `DamageCalculatonType` enum/property to `SpreadDamageWarhead` (the third commit).
_The first two commits are cleanup._

penev92 committed:
Added DamageCalculationType enum to SpreadDamageWarhead

penev92 committed:
Equalized the `DoImpact` methods in damage warheads

penev92 committed:
Reduced HitShape radius of D2k units to minimum

Since they are only used to hold the armor type anyway.

penev92 committed:
Switched D2k to use the new DamageCalculationType

This brings D2k in line with the damage model of the original game.

reaperrr commented on OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs:
```diff
@@ -60,10 +64,28 @@ public override void DoImpact(WPos pos, Actor firedBy, WarheadArgs args)
+				if (falloffDistance > Range[Range.Length - 1].Length)
+					continue;
+
 				var localModifiers = args.DamageModifiers.Append(GetDamageFalloff(closestActiveShape.Second.Length));
```
Not 100% sure, but wouldn't this need to be `GetDamageFalloff(falloffDistance)` then?

penev92 replied:
Derp doesn't even begin to cover it.

penev92 replied:
Fixed.

pchote commented:
Can you please reformat the commit messages to follow the recommendations at https://chris.beams.io/posts/git-commit/ ? Specifically, the subject vs body split and line length limits to keep the messages readable in the github history.

penev92 commented:
Reworded commit messages and rebased for good measure.

abcdefg30 commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -71,6 +71,7 @@ public static IEnumerable<Actor> FindBlockingActorsOnLine(this World world, WPos
 
 		/// <summary>
 		/// Finds all the actors of which their health radius might be intersected by a specified circle.
+		/// Note: Results may include false positives.
```
This alone doesn't seem very helpful. I assume this is because of `world.ActorMap.LargestActorRadius`? Can we add a bit of explanation here?

pchote replied:
This was already implied by the "might" in the sentence above, FWIW.

penev92 replied:
Btw that "might" is making us do lots of unnecessary checks.

pchote replied:
It also saves a significant amount of unnecessary calculations in some important hot paths.

penev92 replied:
My point is we should narrow this down further. I will remove the comment if it's such a problem.

pchote replied:
IMO removing the comment change is the simplest "solution" here.

penev92 replied:
Removed.

penev92 committed:
Cleaned up DamageWarhead

Reordered methods and fixed access modifiers. Also removed unused using statements from warheads.

penev92 commented:
Rebased.

reaperrr reviewed (approved):
My 👍 still applies

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a087e8b1527d2e40c041b061ca13319940ba8e77...26a7c5bef38e3bac5d849fea0c4cc69f79f24718)
